### ISSUE TITLE
HandleInertiaRequests Middleware does not work on fresh app with Jetstream

### ISSUE THREAD

**Issue by ArielMejiaDev** [inertiajs/inertia-laravel#193]
Here the steps:

- Create a new app with jetstream: 

```php
laravel new myapp --jet --stack=inertia --teams
```

- Create a middleware as docs says:

```php
php artisan make:middleware HandleInertiaRequests
```

- add a "share()" method with data that all Laravel apps has:

```php
<?php

namespace App\Http\Middleware;

use Illuminate\Http\Request;
use Inertia\Middleware;

class HandleInertiaRequests extends Middleware
{
    /**
     * The root template that's loaded on the first page visit.
     *
     * @see https://inertiajs.com/server-side-setup#root-template
     * @var string
     */
    protected $rootView = 'app';

    /**
     * Determines the current asset version.
     *
     * @see https://inertiajs.com/asset-versioning
     * @param  \Illuminate\Http\Request  $request
     * @return string|null
     */
    public function version(Request $request)
    {
        return parent::version($request);
    }

    /**
     * Defines the props that are shared by default.
     *
     * @see https://inertiajs.com/shared-data
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    public function share(Request $request)
    {
        return array_merge(parent::share($request), [
            'auth' => function () use ($request) {
                return [
                    'user' => $request->user() ? : null,
                ];
            },
            'flash' => function () use ($request) {
                return [
                    'success' => $request->session()->get('success'),
                    'error' => $request->session()->get('error'),
                ];
            },
        ]);
    }
}

```

- Register the Middleware on "app/Http/Kernel.php" on key "web":

```php
    protected $middlewareGroups = [
        'web' => [
            \App\Http\Middleware\HandleInertiaRequests::class,
        ],
        'api' => [
            'throttle:api',
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],
    ];
```

- Go to vue extension and search on props, (there is no "auth" or "flash" props).

- If I add the Inertia::share methods in "app/Providers/JestreamServiceProvider.php" the "auth" and "flash" props are available to access with:

```vue
$page.props.auth
$page.props.flash
```

**Comment by tincanbox** [comment]
Simple workaround. I don't think this is the best way.  
But if you want to work with the Middleware layer, just do `manually` way.

Middleware
```php
namespace App\Http\Middleware;

use Closure;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Inertia\Inertia;
use Inertia\Middleware;

class HandleInertiaRequestsOrAnyName extends Middleware
{
    // Why $request does not have a typehint?
    function handle($request, Closure $next): Response
    {
        // Yes, this is the `manually` way.
        Inertia::share([

            // Synchronously
            // Don't override predefined properties,
            // like: app currentRouteName errorBags errors jetstream sessions
            'some_info' => [
                'name' => config('app.name')
            ],

            // Lazily
            'user' => (function($request){
                return $request->user() ? $request->user()->only('id', 'name', 'email') : null;
            }),
        ]);

        return parent::handle($request, $next);
    }

    // Middleware::share is not defined... meant be ResponseFactory::share??
    // You can not use codes on the current document.
    public function share(Request $request)
    {
    }
}

```

Template side
```
{{ $page.some_info.name }}
```

**Comment by CryDeTaan** [comment]
I am experiencing the same behaviour when trying to work with shared data, specifically **flash message**. 
Took a similar approach to @ArielMejiaDev but not sure if it is the best. 

I **updated** the `JestreamServiceProvider.php`'s `boot()` function with `Inertia::share([...]);`.

```php
public function boot()
{
    // ...

    Inertia::share([
        'flash' => function (Request $request) {
            return [
                'success' => $request->session()->get('success'),
                'error' => $request->session()->get('error'),
            ];
        },
        'auth' => function (Request $request) {
            return [
                'user' => $request->user() ? [
                    'id' => $request->user()->id,
                    'name' => $request->user()->name
                ] : null,
            ];
        },
    ]);

}

```

Any suggestions/comments on whether this a good approach will be appreciated.

**Comment by claudiodekker** [comment]
Hi everyone,

Apologies for not responding on this earlier. This is indeed a known issue in Jetstream `1.0`, and **has been solved in Jetstream `2.0` which will be released over the next few days**.

That said, because there isn't an _upgrade path_ due to the fact that Jetstream is really just an asset publisher used to 'generate' boilerplating/scaffolding for you, it's still good to have a fix for this.

In essence, @CryDeTaan's fix isn't "bad" per definition, but it's still nice to use Inertia's Middleware for this. 
Luckily, this isn't at all that difficult! All you have to do, is apply the following to your `JetstreamServiceProvider`:

![Screenshot 2020-12-26 at 13 11 13](https://user-images.githubusercontent.com/1752195/103151151-d67d0880-477b-11eb-96db-cd1a95fd596b.png)

Since it's difficult to copy-paste from a screenshot, here's the code mentioned:

```php
use App\Http\Middleware\HandleInertiaRequests;
```

```php
if (class_exists(HandleInertiaRequests::class)) {
    $kernel->appendToMiddlewarePriority(HandleInertiaRequests::class);
}
```

Hope this helps!

**Comment by claudiodekker** [comment]
P.S: One thing I'd like to recommend, is that as of recent `inertiajs/inertia-laravel` installs, it's entirely possible to set (root-level) keys using dot-notation. This is especially nice with Jetstream, as it's already defined some keys, and overriding it means you have to manually merge stuff which isn't that great of an experience.

Basically, something like this should work:
```php
    /**
     * Defines the props that are shared by default.
     *
     * @see https://inertiajs.com/shared-data
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    public function share(Request $request)
    {
        return array_merge(parent::share($request), [
            'auth' => function () use ($request) {
                return [
                    'user' => $request->user() ? : null,
                ];
            },

            // Inlined here, but can also be used anywhere else 
            // For ex: Inside your Inertia::render call, your controllers using Inertia::share etc.
            'auth.user.can.set_nested_keys' => true,
        ]);
    }
```
